lib/src/sql/parser.rs

```diff
@@ -117,12 +131,68 @@ fn locate<'a>(input: &str, tried: &'a str) -> (&'a str, usize, usize) {
 	(tried, 0, 0)
 }
 
+pub(crate) mod depth {
+	use crate::cnf::MAX_COMPUTATION_DEPTH;
+	use crate::sql::Error::ExcessiveDepth;
+	use nom::Err;
+	use std::cell::Cell;
+	use std::thread::panicking;
+
+	thread_local! {
+		/// How many recursion levels deep parsing is currently.
+		static DEPTH: Cell<u8> = Cell::default();
+	}
+
+	/// Call when starting the parser to reset the recursion depth.
+	#[inline(never)]
+	pub(super) fn reset() {
+		DEPTH.with(|cell| {
+			debug_assert_eq!(cell.get(), 0, "previous parsing stopped abruptly");
+			cell.set(0)
+		});
+	}
+
+	/// Call at least once in recursive parsing code paths to limit recursion depth.
+	#[inline(never)]
+	#[must_use = "must store and implicitly drop when returning"]
+	pub(crate) fn dive() -> Result<Diving, Err<crate::sql::Error<&'static str>>> {
+		DEPTH.with(|cell| {
+			let depth = cell.get().saturating_add(4);
+			if depth <= *MAX_COMPUTATION_DEPTH {
+				cell.replace(depth);
+				Ok(Diving)
+			} else {
+				Err(Err::Failure(ExcessiveDepth))
+			}
+		})
+	}
+
+	#[must_use]
+	#[non_exhaustive]
+	pub(crate) struct Diving;
+
+	impl Drop for Diving {
+		fn drop(&mut self) {
+			DEPTH.with(|cell| {
+				if let Some(depth) = cell.get().checked_sub(1) {
```

Comment:
you are adding 4 but subtracting 1. Is it intentional?

Reply:
Thank you for spotting this, it was an error!